File: src/math/detail/KernelImagesForRegion.cc
// -*- LSST-C++ -*-

/*
 * LSST Data Management System
 * Copyright 2008, 2009, 2010 LSST Corporation.
 *
 * This product includes software developed by the
 * LSST Project (http://www.lsst.org/).
 *
 * This program is free software: you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the LSST License Statement and
 * the GNU General Public License along with this program.  If not,
 * see <http://www.lsstcorp.org/LegalNotices/>.
 */

/*
 * Definition of KernelImagesForRegion class declared in detail/ConvolveImage.h
 */
#include <algorithm>
#include <cmath>
#include <iostream>
#include <sstream>
#include <vector>

#include "boost/assign/list_of.hpp"

#include "lsst/pex/exceptions.h"
#include "lsst/log/Log.h"
#include "lsst/afw/geom.h"
#include "lsst/afw/image/ImageUtils.h"
#include "lsst/afw/math/detail/Convolve.h"

namespace pexExcept = lsst::pex::exceptions;
namespace afwGeom = lsst::afw::geom;
namespace afwImage = lsst::afw::image;
namespace mathDetail = lsst::afw::math::detail;

mathDetail::KernelImagesForRegion::KernelImagesForRegion(
        KernelConstPtr kernelPtr,
        lsst::afw::geom::Box2I const &bbox,
        lsst::afw::geom::Point2I const &xy0,
        bool doNormalize)
:
    lsst::daf::base::Citizen(typeid(this)),
    _kernelPtr(kernelPtr),
    _bbox(bbox),
    _xy0(xy0),
    _doNormalize(doNormalize),
    _imagePtrList(4)
{
    if (!_kernelPtr) {
        throw LSST_EXCEPT(pexExcept::InvalidParameterError, "kernelPtr is null");
    }
    LOGL_DEBUG("TRACE5.afw.math.convolve.KernelImagesForRegion",
    "KernelImagesForRegion(bbox(minimum=(%d, %d), extent=(%d, %d)), xy0=(%d, %d), doNormalize=%d, images...)",
       _bbox.getMinX(), _bbox.getMinY(), _bbox.getWidth(), _bbox.getHeight(), _xy0[0], _xy0[1], _doNormalize);
}

mathDetail::KernelImagesForRegion::KernelImagesForRegion(
        KernelConstPtr const kernelPtr,
        lsst::afw::geom::Box2I const &bbox,
        lsst::afw::geom::Point2I const &xy0,
        bool doNormalize,
        ImagePtr bottomLeftImagePtr,
        ImagePtr bottomRightImagePtr,
        ImagePtr topLeftImagePtr,
        ImagePtr topRightImagePtr)
:
    lsst::daf::base::Citizen(typeid(this)),
    _kernelPtr(kernelPtr),
    _bbox(bbox),
    _xy0(xy0),
    _doNormalize(doNormalize),
    _imagePtrList(4)
{
    if (!_kernelPtr) {
        throw LSST_EXCEPT(pexExcept::InvalidParameterError, "kernelPtr is null");
    }
    _insertImage(BOTTOM_LEFT, bottomLeftImagePtr);
    _insertImage(BOTTOM_RIGHT, bottomRightImagePtr);
    _insertImage(TOP_LEFT, topLeftImagePtr);
    _insertImage(TOP_RIGHT, topRightImagePtr);
    LOGL_DEBUG("TRACE5.afw.math.convolve.KernelImagesForRegion",
    "KernelImagesForRegion(bbox(minimum=(%d, %d), extent=(%d, %d)), xy0=(%d, %d), doNormalize=%d, images...)",
       _bbox.getMinX(), _bbox.getMinY(), _bbox.getWidth(), _bbox.getHeight(), _xy0[0], _xy0[1], _doNormalize);
}

mathDetail::KernelImagesForRegion::ImagePtr mathDetail::KernelImagesForRegion::getImage(
        Location location)
const {
    if (_imagePtrList[location]) {
        return _imagePtrList[location];
    }

    ImagePtr imagePtr(new Image(_kernelPtr->getDimensions()));
    _imagePtrList[location] = imagePtr;
    _computeImage(location);
    return imagePtr;
}

lsst::afw::geom::Point2I mathDetail::KernelImagesForRegion::getPixelIndex(
        Location location)
const {
    switch (location) {
        case BOTTOM_LEFT:
            return _bbox.getMin();
            break; // paranoia
        case BOTTOM_RIGHT:
            return afwGeom::Point2I(_bbox.getMaxX() + 1, _bbox.getMinY());
            break; // paranoia
        case TOP_LEFT:
            return afwGeom::Point2I(_bbox.getMinX(), _bbox.getMaxY() + 1);
            break; // paranoia
        case TOP_RIGHT:
            return afwGeom::Point2I(_bbox.getMaxX() + 1, _bbox.getMaxY() + 1);
            break; // paranoia
        default: {
            std::ostringstream os;
            os << "Bug: unhandled location = " << location;
            throw LSST_EXCEPT(pexExcept::InvalidParameterError, os.str());
        }
    }
}

bool mathDetail::KernelImagesForRegion::computeNextRow(
        RowOfKernelImagesForRegion &regionRow)
const {
    if (regionRow.isLastRow()) {
        return false;
    }

    bool hasData = regionRow.hasData();
    int startY;
    if (hasData) {
        startY = regionRow.front()->getBBox().getMaxY() + 1;
    } else {
        startY = this->_bbox.getMinY();
    }

    int yInd = regionRow.incrYInd();
    int remHeight = 1 + this->_bbox.getMaxY() - startY;
    int remYDiv = regionRow.getNY() - yInd;
    int height = _computeNextSubregionLength(remHeight, remYDiv);

    if (hasData) {
        // Move each region up one segment
        bool isFirst = true;
        for (RowOfKernelImagesForRegion::Iterator rgnIter = regionRow.begin(), rgnEnd = regionRow.end();
            rgnIter != rgnEnd; ++rgnIter) {
            (*rgnIter)->_moveUp(isFirst, height);
            isFirst = false;
        }

    } else {
        ImagePtr blImagePtr = getImage(BOTTOM_LEFT);
        ImagePtr brImagePtr;
        ImagePtr tlImagePtr;
        ImagePtr const trImageNullPtr;

        afwGeom::Point2I blCorner = afwGeom::Point2I(this->_bbox.getMinX(), startY);

        int remWidth = this->_bbox.getWidth();
        int remXDiv = regionRow.getNX();
        for (RowOfKernelImagesForRegion::Iterator rgnIter = regionRow.begin(), rgnEnd = regionRow.end();
            rgnIter != rgnEnd; ++rgnIter) {
            int width = _computeNextSubregionLength(remWidth, remXDiv);
            --remXDiv;
            remWidth -= width;

            std::shared_ptr<KernelImagesForRegion> regionPtr(new KernelImagesForRegion(
                _kernelPtr,
                afwGeom::Box2I(blCorner, afwGeom::Extent2I(width, height)),
                _xy0,
                _doNormalize,
                blImagePtr,
                brImagePtr,
                tlImagePtr,
                trImageNullPtr));
            *rgnIter = regionPtr;

            if (!tlImagePtr) {
                regionPtr->getImage(TOP_LEFT);
            }

            blCorner += afwGeom::Extent2I(width, 0);
            blImagePtr = regionPtr->getImage(BOTTOM_RIGHT);
            tlImagePtr = regionPtr->getImage(TOP_RIGHT);
        }
    }
    return true;
}

void mathDetail::KernelImagesForRegion::_computeImage(Location location) const {
    ImagePtr imagePtr = _imagePtrList[location];
    if (!imagePtr) {
        std::ostringstream os;
        os << "Null imagePtr at location " << location;
        throw LSST_EXCEPT(pexExcept::NotFoundError, os.str());
    }

    afwGeom::Point2I pixelIndex = getPixelIndex(location);
    _kernelPtr->computeImage(
        *imagePtr,
        _doNormalize,
        afwImage::indexToPosition(pixelIndex.getX() + _xy0[0]),
        afwImage::indexToPosition(pixelIndex.getY() + _xy0[1]));
}

std::vector<int> mathDetail::KernelImagesForRegion::_computeSubregionLengths(
    int length,
    int nDivisions)
{
    if ((nDivisions > length) || (nDivisions < 1)) {
        std::ostringstream os;
        os << "nDivisions = " << nDivisions << " not in range [1, " << length << " = length]";
        throw LSST_EXCEPT(pexExcept::InvalidParameterError, os.str());
    }
    std::vector<int> regionLengths;
    int remLength = length;
    for (int remNDiv = nDivisions; remNDiv > 0; --remNDiv) {
        int subLength = _computeNextSubregionLength(remLength, remNDiv);
        if (subLength < 1) {
            std::ostringstream os;
            os << "Bug! _computeSubregionLengths(length=" << length << ", nDivisions=" << nDivisions <<
                ") computed sublength = " << subLength << " < 0; remLength = " << remLength;
            throw LSST_EXCEPT(pexExcept::RuntimeError, os.str());
        }
        regionLengths.push_back(subLength);
        remLength -= subLength;
    }
    return regionLengths;
}

void mathDetail::KernelImagesForRegion::_moveUp(
        bool isFirst,
        int newHeight)
{
    // move bbox up (this must be done before recomputing the top kernel images)
    _bbox = afwGeom::Box2I(
        afwGeom::Point2I(_bbox.getMinX(), _bbox.getMaxY() + 1),
        afwGeom::Extent2I(_bbox.getWidth(), newHeight));

    // swap top and bottom image pointers
    _imagePtrList[BOTTOM_RIGHT].swap(_imagePtrList[TOP_RIGHT]);
    _imagePtrList[BOTTOM_LEFT].swap(_imagePtrList[TOP_LEFT]);

    // recompute top right, and if the first image also recompute top left
    _computeImage(TOP_RIGHT);
    if (isFirst) {
        _computeImage(TOP_LEFT);
    }
}


int const mathDetail::KernelImagesForRegion::_MinInterpolationSize = 10;

mathDetail::RowOfKernelImagesForRegion::RowOfKernelImagesForRegion(
        int nx,
        int ny)
:
    _nx(nx),
    _ny(ny),
    _yInd(-1),
    _regionList(nx)
{
    if ((nx < 1) || (ny < 1)) {
        std::ostringstream os;
        os << "nx = " << nx << " and/or ny = " << ny << " < 1";
        throw LSST_EXCEPT(pexExcept::InvalidParameterError, os.str());
    };
}

File: tests/kernelImagesForRegion.py
#!/usr/bin/env python2
from __future__ import absolute_import, division
from __future__ import print_function
from builtins import range

#
# LSST Data Management System
# Copyright 2008, 2009, 2010 LSST Corporation.
#
# This product includes software developed by the
# LSST Project (http://www.lsst.org/).
#
# This program is free software: you can redistribute it and/or modify
# it under the terms of the GNU General Public License as published by
# the Free Software Foundation, either version 3 of the License, or
# (at your option) any later version.
#
# This program is distributed in the hope that it will be useful,
# but WITHOUT ANY WARRANTY; without even the implied warranty of
# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
# GNU General Public License for more details.
#
# You should have received a copy of the LSST License Statement and
# the GNU General Public License along with this program.  If not,
# see <http://www.lsstcorp.org/LegalNotices/>.
#
import math
import unittest

import numpy

import lsst.utils.tests as utilsTests
import lsst.pex.logging as pexLog
import lsst.afw.geom as afwGeom
import lsst.afw.image as afwImage
import lsst.afw.math as afwMath
import lsst.afw.math.detail as mathDetail

VERBOSITY = 0 # increase to see trace

pexLog.Debug("lsst.afw", VERBOSITY)

#-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-

LocNameDict = {
    mathDetail.KernelImagesForRegion.BOTTOM_LEFT: "BOTTOM_LEFT",
    mathDetail.KernelImagesForRegion.BOTTOM_RIGHT: "BOTTOM_RIGHT",
    mathDetail.KernelImagesForRegion.TOP_LEFT: "TOP_LEFT",
    mathDetail.KernelImagesForRegion.TOP_RIGHT: "TOP_RIGHT",
}

NameLocDict = dict((name, loc) for (loc, name) in LocNameDict.items())

class KernelImagesForRegion(utilsTests.TestCase):
    def setUp(self):
        boxCorner = afwGeom.Point2I(11, 50)
        boxExtent = afwGeom.Extent2I(100, 99)
        self.bbox = afwGeom.Box2I(boxCorner, boxExtent)
        self.xy0 = afwGeom.Point2I(100, 251)
        self.kernel = self.makeKernel()

    def tearDown(self):
        self.bbox = None
        self.kernel = None

    def assertRegionCorrect(self, region):
        """Assert that a region has correct corner images

        This test is only relevant for operations that try to reuse the image array data
        """
        regionCopy = mathDetail.KernelImagesForRegion(region.getKernel(), region.getBBox(), region.getXY0(), region.getDoNormalize())

        for location in (
            region.BOTTOM_LEFT,
            region.BOTTOM_RIGHT,
            region.TOP_LEFT,
            region.TOP_RIGHT,
        ):
            actImage = region.getImage(location)
            actImArr = actImage.getArray().transpose().copy()
            desImage = regionCopy.getImage(location)
            desImArr = desImage.getArray().transpose().copy()
            actImArr -= desImArr
            if not numpy.allclose(actImArr, 0):
                actImage.writeFits("actImage%s.fits" % (location,))
                desImage.writeFits("desImage%s.fits" % (location,))
                self.fail("failed on location %s" % (location,))

    def makeKernel(self):
        kCols = 7
        kRows = 6

        # create spatial model
        sFunc = afwMath.PolynomialFunction2D(1)

        minSigma = 0.1
        maxSigma = 3.0

        # spatial parameters are a list of entries, one per kernel parameter;
        # each entry is a list of spatial parameters
        xSlope = (maxSigma - minSigma) / self.bbox.getWidth()
        ySlope = (maxSigma - minSigma) / self.bbox.getHeight()
        xOrigin = minSigma - (self.xy0[0] * xSlope)
        yOrigin = minSigma - (self.xy0[1] * ySlope)
        sParams = (
            (xOrigin, xSlope, 0.0),
            (yOrigin, 0.0,    ySlope),
            (0.0, 0.0, 0.0),
        )

        kFunc =  afwMath.GaussianFunction2D(1.0, 1.0, 0.0)
        kernel = afwMath.AnalyticKernel(kCols, kRows, kFunc, sFunc)
        kernel.setSpatialParameters(sParams)
        return kernel

    def testDoNormalize(self):
        """Test getDoNormalize
        """
        kernel = self.makeKernel()
        for doNormalize in (False, True):
            region = mathDetail.KernelImagesForRegion(kernel, self.bbox, self.xy0, doNormalize)
            self.assert_(region.getDoNormalize() == doNormalize)

    def testGetPixelIndex(self):
        """Test getPixelIndex method
        """
        region = mathDetail.KernelImagesForRegion(self.kernel, self.bbox, self.xy0, False)
        leftInd = self.bbox.getMinX()
        rightInd = self.bbox.getMaxX() + 1
        bottomInd = self.bbox.getMinY()
        topInd = self.bbox.getMaxY() + 1
        int(round((leftInd + rightInd) / 2.0))
        int(round((bottomInd + topInd) / 2.0))

        for location, desIndex in (
            (region.BOTTOM_LEFT,  (leftInd,  bottomInd)),
            (region.BOTTOM_RIGHT, (rightInd, bottomInd)),
            (region.TOP_LEFT,     (leftInd,  topInd)),
            (region.TOP_RIGHT,    (rightInd, topInd)),
        ):
            desPixIndex = afwGeom.Point2I(desIndex[0], desIndex[1])
            self.assert_(region.getPixelIndex(location) == desPixIndex,
                "getPixelIndex(%s) = %s != %s" % (LocNameDict[location], region.getPixelIndex(location),
                    desPixIndex)
            )

    def testComputeNextRow(self):
        """Test computeNextRow method and the resulting RowOfKernelImagesForRegion
        """
        nx = 6
        ny = 5
        regionRow = mathDetail.RowOfKernelImagesForRegion(nx, ny)
        self.assert_(not regionRow.hasData())
        self.assert_(not regionRow.isLastRow())
        self.assert_(regionRow.getYInd() == -1)

        region = mathDetail.KernelImagesForRegion(self.kernel, self.bbox, self.xy0, False)
        floatWidth = self.bbox.getWidth() / float(nx)
        validWidths = (int(math.floor(floatWidth)), int(math.ceil(floatWidth)))
        floatHeight = self.bbox.getHeight() / float(ny)
        validHeights = (int(math.floor(floatHeight)), int(math.ceil(floatHeight)))

        totalHeight = 0
        prevBBox = None
        prevFirstBBox = None
        for yInd in range(ny):
            rowWidth = 0
            isOK = region.computeNextRow(regionRow)
            self.assert_(isOK)
            self.assert_(regionRow.hasData())
            self.assert_(regionRow.isLastRow() == (yInd + 1 >= ny))
            self.assert_(regionRow.getYInd() == yInd)
            firstBBox = regionRow.getRegion(0).getBBox()
            self.assert_(firstBBox.getMinX() == self.bbox.getMinX())
            if yInd == 0:
                self.assert_(firstBBox.getMinY() == self.bbox.getMinY())
            firstBBoxHeight = firstBBox.getHeight()
            self.assert_(firstBBoxHeight in validHeights)
            totalHeight += firstBBoxHeight
            if yInd > 0:
                self.assert_(firstBBox.getMinY() == prevFirstBBox.getMaxY() + 1)
                if yInd == ny - 1:
                    self.assert_(firstBBox.getMaxY() == self.bbox.getMaxY())
            prevFirstBBox = firstBBox
            for xInd in range(nx):
                subregion = regionRow.getRegion(xInd)
                try:
                    self.assertRegionCorrect(subregion)
                except:
                    print("failed on xInd=%s, yInd=%s" % (xInd, yInd))
                    raise
                bbox = subregion.getBBox()
                rowWidth += bbox.getWidth()
                self.assert_(bbox.getWidth() in validWidths)
                self.assert_(bbox.getHeight() == firstBBoxHeight)
                if xInd > 0:
                    self.assert_(bbox.getMinX() == prevBBox.getMaxX() + 1)
                    self.assert_(bbox.getMinY() == prevBBox.getMinY())
                    self.assert_(bbox.getMaxY() == prevBBox.getMaxY())
                    if xInd == nx - 1:
                        self.assert_(bbox.getMaxX() == self.bbox.getMaxX())
                prevBBox = bbox
            self.assert_(rowWidth == self.bbox.getWidth())
        self.assert_(totalHeight == self.bbox.getHeight())
        self.assert_(not region.computeNextRow(regionRow))

    def testExactImages(self):
        """Confirm that kernel image at each location is correct
        """
        desImage = afwImage.ImageD(afwGeom.Extent2I(self.kernel.getWidth(), self.kernel.getHeight()))

        for doNormalize in (False, True):
            region = mathDetail.KernelImagesForRegion(self.kernel, self.bbox, self.xy0, doNormalize)
            for location in (
                region.BOTTOM_LEFT,
                region.BOTTOM_RIGHT,
                region.TOP_LEFT,
                region.TOP_RIGHT,
            ):
                pixelIndex = region.getPixelIndex(location)
                xPos = afwImage.indexToPosition(pixelIndex[0] + self.xy0[0])
                yPos = afwImage.indexToPosition(pixelIndex[1] + self.xy0[1])
                self.kernel.computeImage(desImage, doNormalize, xPos, yPos)

                actImage = region.getImage(location)
                msg = "exact image(%s) incorrect" % (LocNameDict[location],)
                self.assertImagesNearlyEqual(actImage, desImage, msg=msg)


#-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-

def suite():
    """Returns a suite containing all the test cases in this module."""
    utilsTests.init()

    suites = []
    suites += unittest.makeSuite(KernelImagesForRegion)
    suites += unittest.makeSuite(utilsTests.MemoryTestCase)

    return unittest.TestSuite(suites)

def run(doExit=False):
    """Run the tests"""
    utilsTests.run(suite(), doExit)

if __name__ == "__main__":
    run(True)
